lib/ansible/playbook/role/__init__.py

```diff
@@ -556,13 +567,19 @@ def get_all_dependencies(self):
         Returns a list of all deps, built recursively from all child dependencies,
         in the proper order in which they should be executed or evaluated.
         """
-        if self._all_dependencies is None:
+        return self._all_dependencies
 
-            self._all_dependencies = []
-            for dep in self.get_direct_dependencies():
-                for child_dep in dep.get_all_dependencies():
-                    self._all_dependencies.append(child_dep)
-                self._all_dependencies.append(dep)
+    @cached_property
+    def _all_dependencies(self):
+        self._all_dependencies = []
+        for dep in self.get_direct_dependencies():
+            for child_dep in dep._all_dependencies:
+                if child_dep in self._all_dependencies:
+                    self._all_dependencies.remove(child_dep)
+                self._all_dependencies.append(child_dep)
+            if dep in self._all_dependencies:
+                self._all_dependencies.remove(child_dep)
```

Comment:
I suspect this meant to say:
```suggestion
                self._all_dependencies.remove(dep)
```

Otherwise I get:
```
    self._all_dependencies.remove(child_dep)
                                  ^^^^^^^^^
UnboundLocalError: cannot access local variable 'child_dep' where it is not associated with a value
```

Reply:
You're right. I wonder why I didn't catch that before. The code is running for me for several deployments I'm doing 🤔